### ISSUE TITLE
RFC6337 hold-unhold issue

### ISSUE THREAD

**Issue by denyspozniak** [signalwire/freeswitch#846]
**Describe the bug**
When FreeSWITCH is put on hold and after that, a SIP INVITE message arrives without an SDP, then FreeSWITCH does not remove from the hold state.

**To Reproduce**

0. Test case: Leg A -> FS internal profile -> FS external profile -> Leg B
1. Use vanilla config with two profiles (internal and external)
2. Call from internal to external direction
3. Put on hold on external leg B via SIP INVITE with SDP ( a=sendonly ). leg A hears MOH
4. Leg B sends SIP INVITE without SDP towards the external FS side
5. FS sends 200 OK/SDP with a=recvonly
6. Leg B sends ACK/SDP with a=sendrecv
7. Leg A still hears MOH, but should not

If we are using **sip_unhold_nosdp=true**, FreeSWITCH behaves with respect to SIP signaling ( 5. FS sends 200 OK/SDP with a=sendrecv ), but not to media. Leg A continues to hear MOH

**Expected behavior**
As per RFC6337 5.3
```
If a UA has occasion to send another offer in the session, without
any desire to change the hold status (e.g., in response to a re-
INVITE without an offer, or when sending a re-INVITE to refresh the
session timer), it should follow the "General Principle for
Constructing Offers and Answers" (Section 5.1).  If it previously
initiated a "hold" by sending "a=sendonly" attribute or "a=inactive"
attribute, then it should offer that again.  If it had not previously
initiated "hold", then it should offer "a=sendrecv" attribute, even
if it had previously been forced to answer something else.  Without
this behavior it is possible to get "stuck on hold" in some cases,
especially when a 3pcc is involved.

```

**Package version or git hash**
 - FreeSWITCH Version 1.10.2-release.4~64bit (-release.4 64bit)

**Trace logs**
![image](https://user-images.githubusercontent.com/22612345/92527432-a858d480-f22f-11ea-83d6-1a92127e536e.png)

**Comment by melekoktay** [comment]
Hi,

We got same issue in our system. After researching problem , we saw [this commit](https://github.com/signalwire/freeswitch/commit/2e3227b50f897ab7471b09979ba25b6ca0ff5887)  is foundation of our problem. 


[freeswitch.hold.patch.zip](https://github.com/signalwire/freeswitch/files/5193245/freeswitch.hold.patch.zip)
Then we apply following patch to resolve the issue. It works. I wish it helps

**Comment by denyspozniak** [comment]
@melekoktay thanks!
Is the hold state disabled for the originator?

**Comment by melekoktay** [comment]
in some how we can say  yes to this -->  "Is the hold state disabled for the originator?" 
But i think it is BUG, and nobody report it until now. Probably, core developers are not aware !

**Comment by denyspozniak** [comment]
Unfortunately, but it didn't help me.
I have simple dialplan below:
```
  <context name="public">
    <extension name="test">
        <condition field="destination_number" expression="^(\d+)$">
          <action application="info"/>
          <action application="export" data="nolocal:sip_unhold_nosdp=true"/>
          <action application="bridge" data="sofia/default/$1@192.168.11.114:5080"/>
        </condition>
    </extension>
   </context>

```

**Comment by denyspozniak** [comment]
Unfortunately, the proposed patch did not help us.
Dialplan:
```
  <context name="public">
    <extension name="test">
        <condition field="destination_number" expression="^(\d+)$">
          <action application="info"/>
          <action application="export" data="nolocal:sip_unhold_nosdp=true"/>
          <action application="bridge" data="sofia/default/$1@192.168.11.114:5080"/>
        </condition>
    </extension>
   </context>

```

200 OK/SDP does not have **a=recvonly**, but MOH is present on A-leg:

![image](https://user-images.githubusercontent.com/22612345/92648429-11a11c00-f2f2-11ea-89ba-23521c18183d.png)
![image](https://user-images.githubusercontent.com/22612345/92648703-84aa9280-f2f2-11ea-8768-366778af9f2d.png)

**Comment by melekoktay** [comment]
You should add "rtp_renegotiate_codec_on_reinvite=false " also on your dialplan

<action application="set" data="rtp_renegotiate_codec_on_reinvite=false"/>

**Comment by denyspozniak** [comment]
Tested with dialplan below, but no luck:
```
    <extension name="test">
        <condition field="destination_number" expression="^(\d+)$">
            <action application="bridge" data="{sip_unhold_nosdp=true,sip_renegotiate_codec_on_reinvite=false}sofia/default/$1@192.168.11.114:5080"/>
        </condition>
    </extension>
   </context>

```

**Comment by melekoktay** [comment]
you wrote "**sip**_renegotiate_codec_on_reinvite" , but I said "**rtp_**renegotiate_codec_on_reinvite"

**Comment by mjerris** [comment]
> you wrote "sip_renegotiate_codec_on_reinvite" , but I said "rtp_renegotiate_codec_on_reinvite"
> 
> sip.. --> rtp...

neither does anything

**Comment by mjerris** [comment]
sip_unhold_nosdp=true is the fix for this issue.  you may be setting it on the wrong leg.

**Comment by denyspozniak** [comment]
> sip_unhold_nosdp=true is the fix for this issue. you may be setting it on the wrong leg.

@mjerris 
I set this parameter to B-leg where exactly the call is put on hold and removed from this state.
And yes, SIP signaling is perfect, but FreeSWITCH keeps playing MOH on leg A...

**Comment by melekoktay** [comment]
> sip_unhold_nosdp=true is the fix for this issue. you may be setting it on the wrong leg.

does not work neither **v1.6.20** & **v1.10.4** versions

**Comment by ovi82** [comment]
This patch worked for us to fix some Broadworks compatibility issues but was never merged:
https://freeswitch.org/stash/projects/FS/repos/freeswitch/pull-requests/1362/overview

**Comment by jamesfranck** [comment]
> This patch worked for us to fix some Broadworks compatibility issues but was never merged:
> https://freeswitch.org/stash/projects/FS/repos/freeswitch/pull-requests/1362/overview

I don't have access to this issue, can you please share me the fix here pls ?

**Comment by ben45** [comment]
Hi @ovi82, we cannot acces to https://freeswitch.org/stash/projects/FS/repos/freeswitch/pull-requests/1362/overview
Can you please share the fix please ? 
thanks

**Comment by ovi82** [comment]
Please find below the added code and a screenshot of the diff:

			if (switch_channel_var_true(channel, "sip_unhold_nosdp")) {
					ss_state = nua_callstate_completed;
					goto state_process;
				}

It was done on FS 1.6 in 2017 but still seems to be an actual issue.

![Screen Shot 2021-03-17 at 10 46 42](https://user-images.githubusercontent.com/13569330/111447887-28478600-870e-11eb-83a8-6ea40414335d.png)

**Comment by ben45** [comment]
Thanks for fast answer,
We will test to add this code on lab 1.10.5

Regards,

**Comment by bferreirq** [comment]
Hi all,
Is it possible to merged this fix directly in new release ?
Regards,

**Comment by jamesfranck** [comment]
Hello,
Did you have any luck to merge this request ? 
We are facing also the same issue solved by your suggested changes.

Thanks in advance.

**Comment by olofssonanton** [comment]
I found this issue when searching for leads to fix an issue with 3PCC, where the call is never taken off hold after receiving a re-invite without SDP.

This should be reopened, right? As far as I can see media still is not working with sip_unhold_nosdp=true, and without the changes mentioned by @ovi82 there's no way to get this working accordingly to the RFC excerpt mentioned by @denyspozniak.

Also, based on that RFC excerpt, should this perhaps be the default behaviour (i.e. not only when sip_unhold_nosdp=true)?

**Comment by olofssonanton** [comment]
I found two other issues mentioning this RFC deviation.
https://github.com/signalwire/freeswitch/issues/1257
https://github.com/signalwire/freeswitch/issues/1091 - here @shaunjstokes has a PR to remove `sip_unhold_nosdp` and instead always send SDP with a=sendrecv.

@mjerris you seem to have experience in this area - do you have any input on this? Do you know why `sip_unhold_nosdp=true` isn't the default behaviour?

**Comment by olofssonanton** [comment]
Sorry for the unsolicited mention @andywolk, but you seem to be the most active contributor on the FreeSWITCH. I feel like this issue need some more attention but I do not want to create yet another issue, further fragmenting the discussion. Are you familiar with this, or do you know who else might be able to able to shine some light on this?

**Comment by andywolk** [comment]
Please try with the latest FS.

**Comment by olofssonanton** [comment]
Hey @andywolk, I recently had the chance to test this again. It unfortunately seems like the situation is unchanged on FS 1.10.7, meaning that the call is still kept on hold after receiving the re-INVITE with empty SDP (regardless of whether or not `sip_unhold_nosdp=true` is set).

**Comment by tecwerks** [comment]
We ran into this issue today, but adding
`<action application="export" data="nolocal:sip_unhold_nosdp=true" />`
immediately before the bridge statement to our carrier resolved the issue for us in version 1.10.7.

**Comment by denyspozniak** [comment]
@tecwerks Is your FreeSWITCH involved in media stream processing?

**Comment by tecwerks** [comment]
@denyspozniak I've run into this issue again.

There is a pull request #1092 that has yet to be merged, but it will mark sip_unhold_nosdp to always behave as true once it is merged.

I've run into a new situation where calls are picked up by Cisco IVR systems, and begin as a=sendonly, since there is no two way conversation - it's just an IVR. Once the call is picked up by an agent, we receive the Re-INVITE w/o SDP, which FreeSWITCH processes properly as the variable is set, but since the prior state was a=sendonly, that's what it returns to.

This gives the appearance of the variable not working, but because a two-way audio path was never established, it's technically doing what it was told.

To combat this behavior, as well as to hear the remote MOH/on hold message from the far end, and not our local MOH:

`"application="export" data="nolocal:rtp_disable_hold=true"`

You will ignore remote re-invites that attempt to throw you on hold, while retaining your ability to still put calls on hold with the A leg.

The combination of these two variables has fixed my issues.

**Comment by denyspozniak** [comment]
Thanks for the useful information!
Maybe it will be useful for someone: I was able to get around this problem through manipulations with the PSD on Kamailio.

**Comment by denyspozniak** [comment]
Hello!
@chrisonjacob ,
If you are familiar with Kamailio scripting, then the code snippet below will be clear to you.
Here it just happens to manipulate the SIP/SDP part (without modifying the session version).

```
onreply_route[RFC6337_REPLY] {
	route(RFC6337);
}
route[RFC6337] {
	if ( is_reply() ) {
		if ( $rs == 200 && $dlg_var(unhold_from) != $null ) {
			if ( $ua =~ "FreeSWITCH" && $ct =~ "mod_sofia" ) {
				replace_body("a=recvonly", "a=sendrecv");
			}
		}
		return;
	}

	if ( is_method("INVITE") && has_totag() && $si == '192.168.100.100' ) {
		if( !t_is_set("onreply_route") ) t_on_reply("RFC6337_REPLY");

		if ( is_audio_on_hold() && $dlg_var(hold_lock) != 1 ) {
			$dlg_var(hold_from) = $var(from_who);
			$dlg_var(hold_body) = $rb;
			$dlg_var(hold_lock) = "1";
		} else if ( !has_body("application/sdp") && $dlg_var(hold_lock) == "1" ) {
			$dlg_var(unhold_from) = $var(from_who);
			$dlg_var(unhold_cseq) = $cs;
			$dlg_var(hold_lock)   = "0";
			xlog("L_WARN", "[RFC6337] *SDP attached* $rm/fU=$fU/tU=$tU/$si $ci \n");
			$var(hold_body) = $(dlg_var(hold_body){re.subst,/a=sendonly/a=sendrecv/g});
			set_body("$var(hold_body)", "application/sdp");
		}
	}

	if ( is_method("ACK") && has_body("application/sdp") && search_body("a=sendrecv") ) {
		if ( $dlg_var(hold_from) == $dlg_var(unhold_from) && $dlg_var(unhold_cseq) == $cs ) {
			xlog("L_WARN", "[RFC6337] *SDP un-hold detected* $rm/fU=$fU/tU=$tU/$si $ci \n");
			remove_hf("Content-Type");
			remove_body();
			msg_apply_changes();
		}
	}
}
```

**Comment by TechGuyVN** [comment]
> Hello! @chrisonjacob , If you are familiar with Kamailio scripting, then the code snippet below will be clear to you. Here it just happens to manipulate the SIP/SDP part (without modifying the session version).
> 
> ```
> onreply_route[RFC6337_REPLY] {
> 	route(RFC6337);
> }
> route[RFC6337] {
> 	if ( is_reply() ) {
> 		if ( $rs == 200 && $dlg_var(unhold_from) != $null ) {
> 			if ( $ua =~ "FreeSWITCH" && $ct =~ "mod_sofia" ) {
> 				replace_body("a=recvonly", "a=sendrecv");
> 			}
> 		}
> 		return;
> 	}
> 
> 	if ( is_method("INVITE") && has_totag() && $si == '192.168.100.100' ) {
> 		if( !t_is_set("onreply_route") ) t_on_reply("RFC6337_REPLY");
> 
> 		if ( is_audio_on_hold() && $dlg_var(hold_lock) != 1 ) {
> 			$dlg_var(hold_from) = $var(from_who);
> 			$dlg_var(hold_body) = $rb;
> 			$dlg_var(hold_lock) = "1";
> 		} else if ( !has_body("application/sdp") && $dlg_var(hold_lock) == "1" ) {
> 			$dlg_var(unhold_from) = $var(from_who);
> 			$dlg_var(unhold_cseq) = $cs;
> 			$dlg_var(hold_lock)   = "0";
> 			xlog("L_WARN", "[RFC6337] *SDP attached* $rm/fU=$fU/tU=$tU/$si $ci \n");
> 			$var(hold_body) = $(dlg_var(hold_body){re.subst,/a=sendonly/a=sendrecv/g});
> 			set_body("$var(hold_body)", "application/sdp");
> 		}
> 	}
> 
> 	if ( is_method("ACK") && has_body("application/sdp") && search_body("a=sendrecv") ) {
> 		if ( $dlg_var(hold_from) == $dlg_var(unhold_from) && $dlg_var(unhold_cseq) == $cs ) {
> 			xlog("L_WARN", "[RFC6337] *SDP un-hold detected* $rm/fU=$fU/tU=$tU/$si $ci \n");
> 			remove_hf("Content-Type");
> 			remove_body();
> 			msg_apply_changes();
> 		}
> 	}
> }
> ```

anyone try to convert this source to opensips, i see many function not work and don't know where to place the code , pls help
